### PR TITLE
Setup schema and models for API Key authentication

### PR DESCRIPTION
**Pull request by kevinansfield**
refs https://github.com/TryGhost/Ghost/issues/9865
- schema migration
  - adds `integrations` and `api_keys` tables
  - inserts `Admin API Client` role and permissions
- adds `Integration` model
- adds `ApiKey` model
  - creates default secret if not given
  - hardcodes associated role based on key type
    - `admin` = `Admin API Client`
    - `content` = no role

Split out from https://github.com/TryGhost/Ghost/pull/9869 to make review easier.

### CODE REVIEW

**Comment by kirrg001** [comment]
@kevinansfield When do you need a review on this? Let me know if it's blocking you.

**Comment by kevinansfield** [comment]
@kirrg001 as soon as you have a moment would be 👍 I'm splitting more PRs out from #9869 but they all build on top of this one

File: core/server/data/schema/schema.js
Comment (kirrg001) [inline core/server/data/schema/schema.js]: Is this an image path? We use the convention adding `_image` at the end.
`icon_image` sounds a little weird, but maybe there is a better name.
Ignore this if `icon` is not an image path.

Comment (kevinansfield) [reply core/server/data/schema/schema.js]: This is an image path but I didn't realise there was a convention. `icon_image` seems fine, I'll update it 🙂

**Comment by kirrg001** [comment]
I am getting an error when running the migration

```
TypeError: Cannot read property 'name' of undefined
    at Object.addApiKeyRole (/Users/kirrg001/dev/ghost/ghost-workspace2/core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js:39:50)
    at _private.addIntegrationsTable.then.then (/Users/kirrg001/dev/ghost/ghost-workspace2/core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js:79:30)
```

🤷🏻‍♀️

File: core/test/unit/models/api-key_spec.js
Comment (kirrg001) [inline core/test/unit/models/api-key_spec.js]: This does not work, it's not an assertion.
`should.exist(...)`

**Comment by kevinansfield** [comment]
Oops, I pulled the role fixtures out for the tests but missed the relevant schema fixture. Should be fixed now 😅

File: core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js
Comment (kirrg001) [inline core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js]: We separate DML and DDL changes, because DDL changes need a manual rollback logic (down), whereas DML can run a transaction (auto-rollback). Sometimes it's also required to add the down hook for the DML changes, but depends very on the use case.

File: core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js
Comment (kirrg001) [inline core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js]: I am getting

> [2018-09-25 13:12:48] WARN (0/0) Adding permissions_roles fixtures for the admin_api_key role

Is that correct?

File: core/server/models/api-key.js
Comment (kirrg001) [inline core/server/models/api-key.js]: If you don't need any custom logic for adding models, you could just remove `add` here completely. It will automatically call the base model.

File: core/server/models/api-key.js
Comment (kirrg001) [inline core/server/models/api-key.js]: I have not looked at your big PR at all. One question: will we also use `models.ApiKey.edit` or only `add`?
Because setting the role_id to null only happens when you change your type. Just curious what the plan is 👍 
Maybe it's just not part of this PR :)

**Review by kirrg001 — COMMENTED** [review]
I did a first review round. Left some first comments. I need to continue with my tasks, will try to review again later :)

ES6: I know you are used to liberal let 😁but server uses constantly const. If we do two different stylings, it's confusing.

File: core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js
Comment (kevinansfield) [reply core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js]: So this should be split into two separate migration files?

Comment (kirrg001) [reply core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js]: 👍

File: core/server/models/api-key.js
Comment (kevinansfield) [reply core/server/models/api-key.js]: `models.ApiKey.edit` will be used. I assumed `onSaving` is called for both create and update but that assumption may be coming from other ORMs 😬

Does bookshelf call `onSaving` with "hasChanged" for all the properties on `add` or does it need to be handled separately?

Comment (kirrg001) [reply core/server/models/api-key.js]: So yeah, Bookshelf triggers onSaving for insert+update. You could differentiate by using `options.method === insert` or you register `onUpdating` and `onCreating`. That is up to you.

> Does bookshelf call onSaving with "hasChanged" for all the properties on add or does it need to be handled separately?

Ähm not sure i understand 🤪

Comment (kevinansfield) [reply core/server/models/api-key.js]: > So yeah, Bookshelf triggers onSaving for insert+update.

That's good, that's the behaviour I wanted 😄

> Ähm not sure i understand 🤪

When `onSaving` is called with `options.method === insert` will `this.hasChanged('type')` always be true if a type value has been set?

The main reason for the "hasChanged" conditional was to avoid an extra DB call every time an API Key gets saved which could be fairly often if we're storing the `lastSeenAt` value on each API call.

Comment (kirrg001) [reply core/server/models/api-key.js]: > When onSaving is called with options.method === insert will this.hasChanged('type') always be true if a type value has been set?

Exceptions:

For adding: if the user has not set the type and there is also not default value in the model, `hasChanged` will be false.

For editing: if you don't edit `data.type` (previous === current value), it's false.

I hope that helps 😁

Comment (kevinansfield) [reply core/server/models/api-key.js]: 👍 thanks! That matches what I expected. In this case the conditional being false is fine if the user hasn't set a `type` because that will throw a validation error (the type column isn't nullable and has an `isIn` validation)

**Comment by kevinansfield** [comment]
Updated, and ready for review again 😄

File: core/server/models/api-key.js
Comment (kirrg001) [reply core/server/models/api-key.js]: 👍 Can't the user in theory set the same type as stored in the db (previous) and just a different role_id (even for the content type)?

Maybe it's not important for now 😁

Comment (kevinansfield) [reply core/server/models/api-key.js]: Good catch 🎣 We should set the hard coded role id if the type or role has changed. I’ll push a fix later

File: core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js
Comment (allouis) [inline core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js]: You seem to like this functional style of programming and I like it too! 

Usually though I would go with the "data last" principle for fn arguments, taking configuration/options as the first arguments and the thing to be operated on last.

That would enable you to then use a fn sig like `(type, options) => (table) -> { /...` so in your `Promise.each` below you can just do `Promise.each(tables, _private.addOrRemoveTable(type, options))`

This isn't a change request tho so :shrimp: :relaxed:

File: core/server/models/api-key.js
Comment (allouis) [inline core/server/models/api-key.js]: Could you just return here?

File: core/server/models/api-key.js
Comment (allouis) [reply core/server/models/api-key.js]: Out of interest how/why would edit be used? Just for permissions, or would we allow editing the type? I think that could be dangerous - because editing the type from content to admin without updating the secret would allow a previously compromised content key to then be used with the admin api.

File: core/server/models/integration.js
Comment (allouis) [inline core/server/models/integration.js]: Are you planning to add to this empty object?

**Review by allouis — COMMENTED** [review]
Looks good, few minor tiny comments

File: core/server/models/api-key.js
Comment (kevinansfield) [reply core/server/models/api-key.js]: From the UI perspective `edit` would only be used for generating a new secret, however this is the model layer so I think it should be more flexible. Restrictions can be placed at the API layer to limit what can be done externally.

File: core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js
Comment (kevinansfield) [reply core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js]: Maybe it's because I'm not used to seeing it but I find `Promise.each(tables, _private.addOrRemoveTable(type, options))` harder to quickly parse due to the `table` argument being "hidden".

However I do agree that argument order should have more consideration given to it so that it's easier to use functions in a functional style when appropriate. TBH as this was a one-time migration file I didn't give it a lot of thought 😄 The `addOrRemoveTable` function looks like a good candidate for a general util function though so it's a good point to raise.

File: core/server/models/api-key.js
Comment (kevinansfield) [reply core/server/models/api-key.js]: 👍

File: core/server/models/api-key.js
Comment (kevinansfield) [reply core/server/models/api-key.js]: It's "never" right now but I imagine that could change soon when we get to implementing Members.

As a scenario, should Content API keys only ever be for public content? What if I want to pull some members-only content in to my marketing site to show what's available behind the pay-wall - should I need to use a full Admin API key for that or can I create a "private" Content API key that my site can use?

Comment (kevinansfield) [reply core/server/models/api-key.js]: Updated to enforce the hardcoded roles if either the type or role gets changed.

File: core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js
Comment (kevinansfield) [reply core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js]: I tried the more functional style out locally but I found it ended up with harder to follow code, I'll leave as-is for now 😄

File: core/server/models/api-key.js
Comment (allouis) [reply core/server/models/api-key.js]: > As a scenario, should Content API keys only ever be for public content?

Good point, as members (private) content will be on the content api

>  What if I want to pull some members-only content in to my marketing site to show what's available behind the pay-wall - should I need to use a full Admin API key for that or can I create a "private" Content API key that my site can use?

I was under the impression that we'd be making "private"("members") content api keys for that data yeah :)

The issue I have with this though, is not the different types of API keys, it's the ability to change a key from one type to another, or more specifically to change the permissions of an "active" key.

I think if you need a key with different permissions, then you should create a new key, with a new id and secret, that way any previous use of the key in a public context can be guaranteed to not be an issue.

(FWIW I don't think this would be an issue if we didn't suggest that some keys _can_ be used in the browser, if everything was serverside, then changing permissions of "active" keys wouldn't be so much of an issue.)

I don't want this convo to hold up the PR tho so i'm happy to revisit it at a later date :relaxed:

Comment (kevinansfield) [reply core/server/models/api-key.js]: I agree but let's revisit this later - it's working for now and we can lock it down later if necessary once we have a clearer picture of the requirements.

File: core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js
Comment (allouis) [reply core/server/data/migrations/versions/2.2/2-add-integrations-and-api-key-tables.js]: sure :blush:

File: core/server/data/schema/schema.js
Comment (allouis) [inline core/server/data/schema/schema.js]: should this include a `max: 128` validation as well?

**Comment by kevinansfield** [comment]
Rebased and resolved conflicts

File: core/server/data/schema/schema.js
Comment (kevinansfield) [reply core/server/data/schema/schema.js]: Is a longer key necessarily a bad thing? I guess it could be useful to keep the data consistent though

Comment (allouis) [reply core/server/data/schema/schema.js]: yeah i'm not sure either way tbh! was just if we wanted the validation. I guess if we make a longer key it's no biggy

**Review by allouis — APPROVED** [review]

File: core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js
Comment (kirrg001) [inline core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js]: Can we pls put this into the role model?
If i delete a role, the related permissions are getting destroyed. 

`models.Role.destroy(..)`

Model layer takes care to auto-remove relations.

Comment (kevinansfield) [reply core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js]: Do we have a convention for doing this? I had a quick look and the only place I could see where associated records are cleaned up is when a user is destroyed but that [is handled in the API layer](https://github.com/TryGhost/Ghost/blob/master/core/server/api/users.js#L264-L278)

Comment (kirrg001) [reply core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js]: This was discussed in Slack. @kevinansfield Let me know when it's pushed.

Comment (kevinansfield) [reply core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js]: @kirrg001 pushed the cleanup into the role model with unit test. Added as a separate commit for review but can be squashed

File: core/server/models/role.js
Comment (allouis) [inline core/server/models/role.js]: nice this is simple :+1:

File: core/test/unit/models/role_spec.js
Comment (allouis) [inline core/test/unit/models/role_spec.js]: You can remove this as it's handled by `testUtils.setup`

File: core/test/unit/models/role_spec.js
Comment (allouis) [inline core/test/unit/models/role_spec.js]: Is it deffo necessary to teardown the entire db for this test?

**Review by allouis — COMMENTED** [review]
1 comment and 1 question :+1:

File: core/test/unit/models/role_spec.js
Comment (kevinansfield) [reply core/test/unit/models/role_spec.js]: 🤷‍♂️ I was only following the other tests I saw. In previous projects I've worked on the db was always torn for each test otherwise it was very easy to introduce interdependencies that were hard to debug because errors start showing up in unrelated tests

Comment (allouis) [reply core/test/unit/models/role_spec.js]: Yeah fair enough - I don't actually know either :grin:. I've always added new objects to db and removed them after the test to avoid the overhead - but no biggie if we already to like this!

File: core/server/data/schema/fixtures/fixtures.json
Comment (kirrg001) [inline core/server/data/schema/fixtures/fixtures.json]: Where is the conversation about this namimg here? I can't find it anymore :(

Comment (allouis) [reply core/server/data/schema/fixtures/fixtures.json]: https://github.com/TryGhost/Ghost/pull/9915#discussion_r220881756

Comment (kevinansfield) [reply core/server/data/schema/fixtures/fixtures.json]: It was added to the other PR, https://github.com/TryGhost/Ghost/pull/9915#discussion_r220878922

I'm happy to rename it to `Admin Integration`. The thought behind the original `Admin API Client` naming was because it's a role for external clients connecting to the Admin API but seeing as we're calling clients integrations now I think it makes sense to keep the naming consistent.

File: core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js
Comment (kirrg001) [inline core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js]: Shouldn't the administrator role get's it's new permissions? 🤔 

>                "Administrator": {                  
                    "integration": "all",
                    "api_key": "all"

File: core/server/data/schema/fixtures/fixtures.json
Comment (kirrg001) [reply core/server/data/schema/fixtures/fixtures.json]: @allouis What do you think?

File: core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js
Comment (kevinansfield) [reply core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js]: Yep, I'll get on that. Do migrations have tests at all?

Comment (kirrg001) [reply core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js]: no

**Comment by kevinansfield** [comment]
@kirrg001 do you have any insight on the integration test that has started to fail here? https://travis-ci.org/TryGhost/Ghost/jobs/436072806#L4875

It seems as though using `bookshelf-relations` is changing the `toJSON` output after destroy is called. I don't know have enough context to know if that's fine and the test should be fixed or if I need to find a way to get the output to match what the test is expecting.

File: core/server/data/schema/schema.js
Comment (kirrg001) [inline core/server/data/schema/schema.js]: How do you fetch the api keys usually?
I am wondering if e.g. integration_id needs an index.

File: core/test/utils/fixtures/data-generator.js
Comment (kirrg001) [inline core/test/utils/fixtures/data-generator.js]: Is that on purpose?

File: core/server/data/schema/fixtures/fixtures.json
Comment (allouis) [reply core/server/data/schema/fixtures/fixtures.json]: I don't mind too much, but would note that an integration can have both admin and public keys, and "Admin Integration" _kinda_ implies that the integration is only an admin one.

As Kevin said - we were previously going to call them clients, but IIRC you would have had a different client for public and admin access.

Maybe "Admin Integration Key" or "Admin Integration Client" would be totally explicit? Eor update the description so we'll always know what it's for? or just leave it! :shrimp:

File: core/server/data/schema/schema.js
Comment (kevinansfield) [reply core/server/data/schema/schema.js]: The most frequent fetch will be when an API key is used, in that case it will be fetched by it's id.

The only time we'll fetch by `integration_id` is when listing keys in the admin admin area.

File: core/test/utils/fixtures/data-generator.js
Comment (kevinansfield) [reply core/test/utils/fixtures/data-generator.js]: Yes, it was a comment to show what the integration_id will eventually be when it's [set up later on](https://github.com/TryGhost/Ghost/pull/9904/files/df2833c854ee13abf508c4db058746ada53ae841#diff-36ae953f72910a8d0e6a62f7dee1ca74R406)

**Comment by kirrg001** [comment]
> @kirrg001 do you have any insight on the integration test that has started to fail here? https://travis-ci.org/TryGhost/Ghost/jobs/436072806#L4875
> It seems as though using bookshelf-relations is changing the toJSON output after destroy is called. I don't know have enough context to know if that's fine and the test should be fixed or if I need to find a way to get the output to match what the test is expecting.

Bookshelf relation does not change the JSON output, it modifies the model.
e.g. if you edit relations e.g. tags, you expect to get the relations back. I agree, for deleting it's maybe a little unexpected to receive `role.permissions = []`, but it proofs that no more relations exists. The destroy API endpoints return `null` anyway. Soo i would ignore this for now, because it's an internal behaviour?

File: core/server/data/schema/schema.js
Comment (kirrg001) [reply core/server/data/schema/schema.js]: Okay 👍 Then it's probably fine as is.

File: core/test/utils/fixtures/data-generator.js
Comment (kirrg001) [reply core/test/utils/fixtures/data-generator.js]: Ok 👍

**Comment by kevinansfield** [comment]
> Soo i would ignore this for now, because it's an internal behaviour?

👍 ok, I'll update the test to match the internal behaviour we expect in this case. Thanks!

File: core/server/data/schema/fixtures/fixtures.json
Comment (kirrg001) [reply core/server/data/schema/fixtures/fixtures.json]: > and "Admin Integration" kinda implies that the integration is only an admin one

Hmm it's the role for the **admin** integration. The content integration key does not have a role id. Otherwise the role could be called "Integration"?

I def. find it weird having the word "API" inside the role name. Any role is for the API and uses the API. And is also basically a client. I am an editor and i use the API. My role name is "editor" and not "editor client api".

Comment (allouis) [reply core/server/data/schema/fixtures/fixtures.json]: > Hmm it's the role for the admin integration.

It's more like the role for the admin key of the integration. Like we have "Integration X" and that integration has a content key and an admin key.

Seems this is just semantics though at this point :man_shrugging: I'm happy with either :relaxed:

Comment (kevinansfield) [reply core/server/data/schema/fixtures/fixtures.json]: `Admin Integration` makes the most sense to me at this point and is less ambiguous than `Admin API Client` when placed in context of the general "Integrations" naming.

We may have a `Content Integration` role later once it's possible that members-only content could be accessible via the Content API but there are still quite a few questions around that. Using `Admin Integration` keeps the possibilities open and more clearly defines the level of access that it allows.

Comment (kirrg001) [reply core/server/data/schema/fixtures/fixtures.json]: 👍

File: core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js
Comment (kevinansfield) [reply core/server/data/migrations/versions/2.2/3-insert-admin-api-client-role.js]: Pushed this in a separate migration

**Comment by kirrg001** [comment]
@kevinansfield Pls let me know when the PR is ready for final test

**Comment by kevinansfield** [comment]
@kirrg001 good for final test 🙂

File: core/server/data/schema/fixtures/fixtures.json
Comment (kirrg001) [inline core/server/data/schema/fixtures/fixtures.json]: One last question to the permissions. Not sure if i have asked this already.
This role has a lot of permissions. Is that correct? 🤔

Comment (kevinansfield) [reply core/server/data/schema/fixtures/fixtures.json]: Yes, it should be the same access as Administrator except locked down a little more so that it can't be used to give itself more roles/permissions, alter permissions of any other user, or download/drop the database.

Comment (kirrg001) [reply core/server/data/schema/fixtures/fixtures.json]: > download/drop the database.

👍 Yeah i saw that this was already removed, which is nice 👍 

--

What about:

delete a post? 🤔 
delete a user? isn't it weird that an integration can delete editors+authors?

I am just double checking with you :)

Comment (kevinansfield) [reply core/server/data/schema/fixtures/fixtures.json]: We were thinking how you'd want to use the API if you wanted to integrate an external auth system, or if you are pulling content in from somewhere and need to create new authors on the fly.

The original discussion was that an integration would be tied to a user with Administrator privileges until we realised that would likely give them a little too much power and also cause issues if users are demoted or deleted. With the role we wanted the same level of access but just restricted a little so that credentials can't be leaked. There's another PR coming later that deals with the per-model `permissible` functions that lock things down a bit further.

Comment (kirrg001) [reply core/server/data/schema/fixtures/fixtures.json]: Yeah all of this makes sense, i am just wondering about some particular restrictions.

> There's another PR coming later that deals with the per-model permissible functions that lock things down a bit further.

Is there a branch already?

Comment (kevinansfield) [reply core/server/data/schema/fixtures/fixtures.json]: There is code in https://github.com/TryGhost/Ghost/pull/9869 waiting to be extracted, there were quite a lot of changes in this PR and the authentication PR though so I was holding off until the they're closer to being merged.

Comment (kirrg001) [reply core/server/data/schema/fixtures/fixtures.json]: Okay it makes sense to extend some of the permissible model functions for specific checks. 

But we want to keep the permission for e.g. delete post or delete user for the integration role?

Comment (kevinansfield) [reply core/server/data/schema/fixtures/fixtures.json]: Yes, so far the spec has been been integrations having Administrator-level access.

Comment (kirrg001) [reply core/server/data/schema/fixtures/fixtures.json]: Thanks. If that was confirmed, i am fine with leaving as is :)

**Review by kirrg001 — COMMENTED** [review]
@kevinansfield Feel free to self merge 👍

**Pull request merged**
